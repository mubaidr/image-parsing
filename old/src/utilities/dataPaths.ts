import os from 'os'
import path from 'path'

const isDev: boolean = process.env.NODE_ENV === 'development'
const tmp: string = isDev ? path.resolve('.tmp') : os.tmpdir()
const home: string = path.join(os.homedir(), 'desktop')

const dataPaths = {
  design: path.resolve('__tests__', '_data', 'design.qr.svg'),
  designBarcode: path.resolve('__tests__', '_data', 'design.svg'),
  home,
  images: path.resolve('__tests__', '_data', 'images-qrcode'),
  imagesBarcode: path.resolve('__tests__', '_data', 'images-barcode'),
  key: path.resolve('__tests__', '_data', 'key.xlsx'),
  keyImage: path.resolve('__tests__', '_data', 'key.jpg'),
  questionsModel: isDev
    ? path.resolve('src/data/questions-model.json')
    : path.resolve('dist/data/questions-model.json'),
  result: path.resolve('__tests__', '_data', 'result.xlsx'),
  resultCompiled: path.resolve('__tests__', '_data', 'resultCompiled.xlsx'),
  root: path.resolve('.'),
  temp: tmp,
  tmp,
}

export { dataPaths }
