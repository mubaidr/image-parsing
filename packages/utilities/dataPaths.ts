import os from 'os';
import path from 'path';

const tmp = os.tmpdir();

export const DataPaths = {
  design: path.resolve('tests', '_data', 'design.qr.svg'),
  designBarcode: path.resolve('tests', '_data', 'design.svg'),
  home: os.homedir(),
  images: path.resolve('tests', '_data', 'images-qrcode'),
  imagesBarcode: path.resolve('tests', '_data', 'images-barcode'),
  key: path.resolve('tests', '_data', 'key.xlsx'),
  keyImage: path.resolve('tests', '_data', 'key.jpg'),
  result: path.resolve('tests', '_data', 'result.xlsx'),
  resultCompiled: path.resolve('tests', '_data', 'resultCompiled.xlsx'),
  root: path.resolve('.'),
  temp: tmp,
  tmp,
};
