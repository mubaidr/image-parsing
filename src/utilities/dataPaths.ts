import os from 'os'
import path from 'path'

const isDev: boolean = process.env.NODE_ENV === 'development'

const tmp: string = isDev ? path.resolve('.tmp') : os.tmpdir()
const home: string = path.join(os.homedir(), 'desktop')

const dataPaths = {
  design: path.resolve('_test_data', 'design.qr.svg'),
  designBarcode: path.resolve('_test_data', 'design.svg'),
  home,
  images: path.resolve('_test_data', 'images-qrcode'),
  imagesBarcode: path.resolve('_test_data', 'images-barcode'),
  key: path.resolve('_test_data', 'key.xlsx'),
  keyImage: path.resolve('_test_data', 'key.jpg'),
  questionsModel: path.resolve('src/data/questions-model.json'),
  result: path.resolve('_test_data', 'result.xlsx'),
  root: path.resolve('.'),
  temp: tmp,
  tmp,
}

export { dataPaths }
