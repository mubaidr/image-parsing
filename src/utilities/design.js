const cheerio = require('cheerio')
const fs = require('fs')

const OPTIONS = require('./question-options')

/**
 * Extracts position & dimensions of objects from SVG design File
 *
 * @param {String} designFilePath Path to the svg design file
 * @returns {Object} JSON object
 */
async function getDesignData(dir) {
  const designData = {
    questions: {},
  }
  const ROLL_NO_PATTERN = new RegExp(/rollnobarcode/gi)
  const QUESTION_PATTERN = new RegExp(/(q[1-9][0-9]?[ad])\b/gi)

  const $ = cheerio.load(fs.readFileSync(dir, 'utf8'))
  const svgViewBox = $('svg')[0].attribs.viewBox.split(' ')

  designData.width = Math.ceil(svgViewBox[2] - svgViewBox[0])
  designData.height = Math.ceil(svgViewBox[3] - svgViewBox[1])

  let gotRollNo = false
  let x
  let y
  let rx
  let ry
  let width
  let height
  let left
  let top

  const groups = $('g')
  for (let i = 0; i < groups.length; i += 1) {
    const group = groups[i]

    const title =
      $(group)
        .find('title')
        .first()
        .html()
        .trim()
        .toUpperCase() || ''

    if (!title) continue

    const isQuestionGroup = QUESTION_PATTERN.test(title)
    const isRollNoGroup =
      isQuestionGroup || gotRollNo ? false : ROLL_NO_PATTERN.test(title)

    if (!(isQuestionGroup || isRollNoGroup)) continue

    const transform = $(group)
      .attr('transform')
      .replace(/(translate)|\(|\)/gi, '')
      .split(',')
      .map(val => parseInt(val, 10) || 0)

    const rect = $(group)
      .find('rect')
      .first()

    left = parseInt(rect.attr('x'), 10) || 0
    top = parseInt(rect.attr('y'), 10) || 0

    rx = parseInt(rect.attr('rx'), 10) || 0
    ry = parseInt(rect.attr('ry'), 10) || 0

    x = left - rx + transform[0]
    y = top - ry + transform[1]

    width = parseInt(rect.attr('width'), 10) || 0 + rx
    height = parseInt(rect.attr('height'), 10) || 0 + ry

    if (isQuestionGroup) {
      const optionTitle = title.slice(-1)
      const questionNumber = title.slice(0, -1)

      if (!designData.questions[questionNumber]) {
        designData.questions[questionNumber] = {}
      }

      if (optionTitle === OPTIONS.A) {
        designData.questions[questionNumber].x1 = x - 1
        designData.questions[questionNumber].y1 = y - 1
      } else {
        // donot check last option to suport any number of options
        designData.questions[questionNumber].x2 = x + width + 1
        designData.questions[questionNumber].y2 = y + height + 1
      }
    } else if (isRollNoGroup) {
      designData.rollNo = { x1: x, y1: y, x2: x + width, y2: y + height }
      gotRollNo = true
    }
  }

  return designData
}

module.exports = {
  getDesignData,
}
