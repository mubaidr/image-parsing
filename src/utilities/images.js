const javascriptBarcodeReader = require('javascript-barcode-reader')
const javascriptQRReader = require('jsqr').default
const fastGlob = require('fast-glob')
const path = require('path')
const uuid = require('uuid')
const sharp = require('sharp')

const { IMAGES, NATIVE_IMAGES } = require('./valid-types')
const DATAPATHS = require('./data-paths')

const CACHE = {}

function getSharpObjectFromSource(src) {
  return sharp(src)
  // .jpeg()
  // .flatten()
}

async function convertImage(src) {
  // native supported images
  if (NATIVE_IMAGES.includes(src.split('.').pop())) return src

  // check cache
  const cached = CACHE[src]
  if (cached) return cached

  // generate random tmp url
  const url = path.join(DATAPATHS.tmp, `${uuid()}.jpg`)
  CACHE[src] = url

  // save file for preview
  await getSharpObjectFromSource(src).toFile(url)

  // returns new url
  return url
}

/**
 * Logs provided image data to .tmp folder
 * @param {sharp || String} img Sharp instance or image source
 * @param {String} name Name of file
 */
async function logImageData(src, name) {
  let img

  if (typeof src === 'string') {
    img = getSharpObjectFromSource(src)
  } else {
    img = src.clone()
  }

  return img.jpeg().toFile(path.join(DATAPATHS.tmp, `${name || uuid()}.jpg`))
}

/**
 * Return a list of valid image format files from the provided path
 *
 * @param {String} path Path to sarch for images
 * @param {Array.<String>} format Array of extensions of valid image formats
 * @returns {Array.<String>} List of file paths
 */
function getImagePaths(dir) {
  return fastGlob(`${dir}/*.{${IMAGES.join(',')}}`, { onlyFiles: true })
}

/**
 *
 *
 * @param {Object} designData A JSON Object containing information about the position, width, height of elements in svg design file (available from utiltities/getDesignData)
 * @param {String} path Path of scanned image file
 * @returns {Object} Roll Number
 */
async function getRollNoFromImage(designData, img, isBarcode) {
  const metadata = await img.metadata()
  const rollNoPos = designData.rollNo
  const ratio = metadata.width / designData.width
  const width = Math.ceil((rollNoPos.x2 - rollNoPos.x1) * ratio)
  const height = Math.ceil((rollNoPos.y2 - rollNoPos.y1) * ratio)
  const obj = {
    id: uuid(),
  }

  img
    .extract({
      left: Math.floor(rollNoPos.x1 * ratio),
      top: Math.floor(rollNoPos.y1 * ratio),
      width,
      height,
    })
    .ensureAlpha()

  // debug image
  // logImageData(img)

  const data = await img.toBuffer()

  try {
    let rollNo

    if (isBarcode) {
      rollNo = await javascriptBarcodeReader(
        { data, width, height },
        { barcode: 'code-39' }
      )
    } else {
      rollNo = javascriptQRReader(data, width, height, {
        inversionAttempts: 'dontInvert',
      }).data
    }

    obj.rollNo = rollNo
    obj.hasValidRollNo = !!obj.rollNo
  } catch (err) {
    // console.error(err)

    obj.rollNo = null
    obj.hasValidRollNo = false
  }

  return obj
}

module.exports = {
  convertImage,
  getImagePaths,
  logImageData,
  getRollNoFromImage,
  getSharpObjectFromSource,
}
