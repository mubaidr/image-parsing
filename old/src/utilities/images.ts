import fastGlob from 'fast-glob'
import path from 'path'
import sharp, { Sharp } from 'sharp'
import uuid from 'uuid'

import ImageNativeTypesEnum from './@enums/ImageNativeTypesEnum'
import ImageTypesEnum from './@enums/ImageTypesEnum'
import { cache } from './cache'
import { dataPaths } from './dataPaths'

const getSharpObjectFromSource = (src: string): Sharp => {
  return sharp(src)
    .raw()
    .flatten()
}

const convertImage = async (src: string): Promise<string> => {
  if (!src) {
    throw new Error('Invalid source provided')
  }

  const ext = src.split('.').pop()

  // native supported images
  if (ext && ext in ImageNativeTypesEnum) {
    return src
  }

  // check cache
  const cached = cache.get(src)
  if (cached) {
    return cached
  }

  // generate random tmp url
  const url = path.join(dataPaths.tmp, `${uuid()}.jpg`)
  cache.set(src, url)

  // save file for preview
  await getSharpObjectFromSource(src)
    .jpeg()
    .toFile(url)

  // returns new url
  return url
}

const logImageData = async (
  src: string | Sharp,
  name?: string
): Promise<string> => {
  let img: Sharp
  const target = path.join(dataPaths.tmp, `${name || uuid()}.jpg`)

  if (typeof src === 'string') {
    img = getSharpObjectFromSource(src)
  } else {
    img = src.clone()
  }

  await img.jpeg().toFile(target)

  return target
}

const getImagePaths = async (dir: string): Promise<string[]> => {
  const loc = dir.replace(/\\/g, '/')
  const exts = Object.keys(ImageTypesEnum)
  // avoid deep directory scan in test env
  const glob =
    process.env.NODE_ENV === 'test'
      ? `${loc}/*.{${exts}}`.replace('//', '/')
      : `${loc}/**/*.{${exts}}`.replace('//', '/')

  return fastGlob(glob, {
    absolute: true,
    onlyFiles: true,
  })
}

export { convertImage, getImagePaths, logImageData, getSharpObjectFromSource }
