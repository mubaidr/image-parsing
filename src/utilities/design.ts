import { parse } from 'fast-xml-parser'
import { readFileSync } from 'fs'
import { DesignData, ItemInfo } from './@classes/WorkerManager'
import RegExpPatterns from './@enums/RegExpPatterns'
import { dataPaths } from './dataPaths'

function getDesignPathByID(id: string) {
  //TODO: implement internal store for designs with import/export functionality
  console.log(id)

  return dataPaths.designBarcode
}

export async function getDesignData(id: string): Promise<DesignData> {
  const filePath = getDesignPathByID(id)

  const { svg } = parse(readFileSync(filePath).toString(), {
    attributeNamePrefix: '',
    ignoreAttributes: false,
    parseNodeValue: true,
    parseAttributeValue: true,
    allowBooleanAttributes: true,
  })

  // get width, height from viewbox
  const [x1, y1, x2, y2] = svg.viewBox
    .split(' ')
    .map((i: string) => parseInt(i, 10))
  const svgWidth = x2 - x1
  const svgHeight = y2 - y1

  // prepare pattern matching reg expressions
  const PATTERN_OPTION = new RegExp(RegExpPatterns.OPTION, 'i')
  const PATTERN_BARCODE = new RegExp(RegExpPatterns.BARCODE, 'i')
  const PATTERN_QRCODE = new RegExp(RegExpPatterns.QRCODE, 'i')

  // for export
  let code: ItemInfo = { x: 0, y: 0, width: 0, height: 0 }
  const questions: {
    [key: string]: ItemInfo
  } = {}

  svg.g.forEach((group: any) => {
    const { title, transform, rect } = group

    if (!rect) return

    const { rx, ry } = rect
    let { x, y, width, height } = rect
    let xTransform = 0
    let yTransform = 0

    if (transform) {
      ;[xTransform, yTransform] = transform
        .replace(/(translate)|\(|\)/gi, '')
        .split(',')
        .map((val: string) => parseInt(val, 10) || 0)
    }

    x = Math.floor(x - rx + xTransform - 3)
    y = Math.floor(y - ry + yTransform - 3)
    width = Math.ceil(width + rx + 6)
    height = Math.ceil(height + ry + 6)

    if (PATTERN_OPTION.test(title)) {
      const questionNumber: string = title.slice(0, -1)

      if (!questions[questionNumber]) {
        questions[questionNumber] = {
          x: svgWidth,
          y: svgHeight,
          width: 0,
          height: 0,
        }
      }

      questions[questionNumber].width += width

      if (questions[questionNumber].x > x) {
        questions[questionNumber].x = x
      }

      if (questions[questionNumber].y > y) {
        questions[questionNumber].y = y
      }

      if (questions[questionNumber].width < width) {
        questions[questionNumber].width = width
      }

      if (questions[questionNumber].height < height) {
        questions[questionNumber].height = height
      }
    } else if (PATTERN_BARCODE.test(title) || PATTERN_QRCODE.test(title)) {
      code = { x, y, width, height }
    }
  })

  return {
    code,
    questions,
    width: svgWidth,
    height: svgHeight,
  }
}
