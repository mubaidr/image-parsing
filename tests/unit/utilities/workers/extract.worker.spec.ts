import { dataPaths } from '@/utilities/dataPaths'
import { getDesignData } from '@/utilities/design'
import { getImagePaths } from '@/utilities/images'
import { start } from '@/utilities/workers/extract.worker'

describe('workerExtract', () => {
  test('should be able to extract key just fine', async () => {
    const designData = await getDesignData(dataPaths.designBarcode)
    const results = await start(
      {
        designData,
        imagePaths: [dataPaths.keyImage],
      },
      false
    )

    expect(results).toBeDefined()
    expect(results?.length).toBe(1)
    // expect(results).toMatchSnapshot()
  })

  test('should be able to extract result just fine', async () => {
    const designData = await getDesignData(dataPaths.designBarcode)
    const imagePaths = await getImagePaths(dataPaths.imagesBarcode)

    const results = await start(
      {
        designData,
        imagePaths,
      },
      false
    )

    expect(results).toBeDefined()
    expect(results?.length).toBeGreaterThanOrEqual(3)
  })

  test('extracted result should match with results.xlsx', async () => {
    // read excel file
    // extract results
    // compare using file path (to  compare answer sheet without roll no)
  })
})
