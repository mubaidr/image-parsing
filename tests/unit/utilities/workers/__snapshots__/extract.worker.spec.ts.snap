// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`workerExtract should be able to extract key just fine 1`] = `
Object {
  "answers": Object {
    "q1": Object {
      "value": "d",
    },
    "q10": Object {
      "value": "c",
    },
    "q11": Object {
      "value": "b",
    },
    "q12": Object {
      "value": "d",
    },
    "q13": Object {
      "value": "b",
    },
    "q14": Object {
      "value": "a",
    },
    "q15": Object {
      "value": "d",
    },
    "q16": Object {
      "value": "c",
    },
    "q17": Object {
      "value": "b",
    },
    "q18": Object {
      "value": "a",
    },
    "q19": Object {
      "value": "c",
    },
    "q2": Object {
      "value": "b",
    },
    "q20": Object {
      "value": "b",
    },
    "q21": Object {
      "value": "a",
    },
    "q22": Object {
      "value": "b",
    },
    "q23": Object {
      "value": "a",
    },
    "q24": Object {
      "value": "b",
    },
    "q25": Object {
      "value": "a",
    },
    "q26": Object {
      "value": "b",
    },
    "q27": Object {
      "value": "a",
    },
    "q28": Object {
      "value": "b",
    },
    "q29": Object {
      "value": "d",
    },
    "q3": Object {
      "value": "d",
    },
    "q30": Object {
      "value": "a",
    },
    "q31": Object {
      "value": "b",
    },
    "q32": Object {
      "value": "a",
    },
    "q33": Object {
      "value": "b",
    },
    "q34": Object {
      "value": "b",
    },
    "q35": Object {
      "value": "a",
    },
    "q36": Object {
      "value": "c",
    },
    "q37": Object {
      "value": "b",
    },
    "q38": Object {
      "value": "a",
    },
    "q39": Object {
      "value": "a",
    },
    "q4": Object {
      "value": "a",
    },
    "q40": Object {
      "value": "a",
    },
    "q41": Object {
      "value": "a",
    },
    "q42": Object {
      "value": "c",
    },
    "q43": Object {
      "value": "a",
    },
    "q44": Object {
      "value": "c",
    },
    "q45": Object {
      "value": "b",
    },
    "q46": Object {
      "value": "a",
    },
    "q47": Object {
      "value": "c",
    },
    "q48": Object {
      "value": "b",
    },
    "q49": Object {
      "value": "a",
    },
    "q5": Object {
      "value": "c",
    },
    "q50": Object {
      "value": "d",
    },
    "q51": Object {
      "value": "a",
    },
    "q52": Object {
      "value": "b",
    },
    "q53": Object {
      "value": "b",
    },
    "q54": Object {
      "value": "d",
    },
    "q55": Object {
      "value": "b",
    },
    "q56": Object {
      "value": "b",
    },
    "q57": Object {
      "value": "a",
    },
    "q58": Object {
      "value": "d",
    },
    "q59": Object {
      "value": "a",
    },
    "q6": Object {
      "value": "d",
    },
    "q60": Object {
      "value": "a",
    },
    "q7": Object {
      "value": "c",
    },
    "q8": Object {
      "value": "a",
    },
    "q9": Object {
      "value": "d",
    },
  },
  "correctCount": 0,
  "id": Any<String>,
  "imageFile": Any<String>,
  "incorrectCount": 0,
  "isCompiled": false,
  "isRollNoExtracted": false,
  "marks": 0,
  "post": "",
  "questionPaperType": "",
  "rollNo": undefined,
  "skippedCount": 0,
  "testCenter": "",
  "testTime": "",
  "totalMarks": 0,
  "unattemptedCount": 0,
}
`;

exports[`workerExtract should be able to extract result just fine 1`] = `
Object {
  "answers": Object {
    "q1": Object {
      "value": "d",
    },
    "q10": Object {
      "value": "d",
    },
    "q11": Object {
      "value": "c",
    },
    "q12": Object {
      "value": "d",
    },
    "q13": Object {
      "value": "c",
    },
    "q14": Object {
      "value": "b",
    },
    "q15": Object {
      "value": "a",
    },
    "q16": Object {
      "value": "b",
    },
    "q17": Object {
      "value": "c",
    },
    "q18": Object {
      "value": "b",
    },
    "q19": Object {
      "value": "c",
    },
    "q2": Object {
      "value": "d",
    },
    "q20": Object {
      "value": "a",
    },
    "q21": Object {
      "value": "d",
    },
    "q22": Object {
      "value": "d",
    },
    "q23": Object {
      "value": "?",
    },
    "q24": Object {
      "value": "d",
    },
    "q25": Object {
      "value": "d",
    },
    "q26": Object {
      "value": "c",
    },
    "q27": Object {
      "value": "c",
    },
    "q28": Object {
      "value": "a",
    },
    "q29": Object {
      "value": "b",
    },
    "q3": Object {
      "value": "c",
    },
    "q30": Object {
      "value": "c",
    },
    "q31": Object {
      "value": "b",
    },
    "q32": Object {
      "value": "a",
    },
    "q33": Object {
      "value": "d",
    },
    "q34": Object {
      "value": "b",
    },
    "q35": Object {
      "value": "a",
    },
    "q36": Object {
      "value": "b",
    },
    "q37": Object {
      "value": "c",
    },
    "q38": Object {
      "value": "d",
    },
    "q39": Object {
      "value": "a",
    },
    "q4": Object {
      "value": "a",
    },
    "q40": Object {
      "value": "b",
    },
    "q41": Object {
      "value": "c",
    },
    "q42": Object {
      "value": "a",
    },
    "q43": Object {
      "value": "a",
    },
    "q44": Object {
      "value": "c",
    },
    "q45": Object {
      "value": "b",
    },
    "q46": Object {
      "value": "d",
    },
    "q47": Object {
      "value": "a",
    },
    "q48": Object {
      "value": "c",
    },
    "q49": Object {
      "value": "a",
    },
    "q5": Object {
      "value": "c",
    },
    "q50": Object {
      "value": "d",
    },
    "q51": Object {
      "value": "d",
    },
    "q52": Object {
      "value": "b",
    },
    "q53": Object {
      "value": "b",
    },
    "q54": Object {
      "value": "d",
    },
    "q55": Object {
      "value": "d",
    },
    "q56": Object {
      "value": "d",
    },
    "q57": Object {
      "value": "b",
    },
    "q58": Object {
      "value": "b",
    },
    "q59": Object {
      "value": "d",
    },
    "q6": Object {
      "value": "d",
    },
    "q60": Object {
      "value": "d",
    },
    "q7": Object {
      "value": "c",
    },
    "q8": Object {
      "value": "a",
    },
    "q9": Object {
      "value": "c",
    },
  },
  "correctCount": 0,
  "id": Any<String>,
  "imageFile": Any<String>,
  "incorrectCount": 0,
  "isCompiled": false,
  "isRollNoExtracted": true,
  "marks": 0,
  "post": "",
  "questionPaperType": "",
  "rollNo": "10023",
  "skippedCount": 0,
  "testCenter": "",
  "testTime": "",
  "totalMarks": 0,
  "unattemptedCount": 0,
}
`;

exports[`workerExtract should be able to extract result just fine 2`] = `
Object {
  "answers": Object {
    "q1": Object {
      "value": "d",
    },
    "q10": Object {
      "value": "?",
    },
    "q11": Object {
      "value": "b",
    },
    "q12": Object {
      "value": "d",
    },
    "q13": Object {
      "value": "b",
    },
    "q14": Object {
      "value": "a",
    },
    "q15": Object {
      "value": "d",
    },
    "q16": Object {
      "value": "c",
    },
    "q17": Object {
      "value": "b",
    },
    "q18": Object {
      "value": "a",
    },
    "q19": Object {
      "value": "c",
    },
    "q2": Object {
      "value": "b",
    },
    "q20": Object {
      "value": "b",
    },
    "q21": Object {
      "value": "a",
    },
    "q22": Object {
      "value": "*",
    },
    "q23": Object {
      "value": "a",
    },
    "q24": Object {
      "value": "?",
    },
    "q25": Object {
      "value": "a",
    },
    "q26": Object {
      "value": "b",
    },
    "q27": Object {
      "value": "a",
    },
    "q28": Object {
      "value": "b",
    },
    "q29": Object {
      "value": "d",
    },
    "q3": Object {
      "value": "d",
    },
    "q30": Object {
      "value": "a",
    },
    "q31": Object {
      "value": "b",
    },
    "q32": Object {
      "value": "a",
    },
    "q33": Object {
      "value": "b",
    },
    "q34": Object {
      "value": "b",
    },
    "q35": Object {
      "value": "a",
    },
    "q36": Object {
      "value": "c",
    },
    "q37": Object {
      "value": "?",
    },
    "q38": Object {
      "value": "a",
    },
    "q39": Object {
      "value": "a",
    },
    "q4": Object {
      "value": "a",
    },
    "q40": Object {
      "value": "a",
    },
    "q41": Object {
      "value": "a",
    },
    "q42": Object {
      "value": "?",
    },
    "q43": Object {
      "value": "a",
    },
    "q44": Object {
      "value": "c",
    },
    "q45": Object {
      "value": "b",
    },
    "q46": Object {
      "value": "a",
    },
    "q47": Object {
      "value": "?",
    },
    "q48": Object {
      "value": "b",
    },
    "q49": Object {
      "value": "a",
    },
    "q5": Object {
      "value": "c",
    },
    "q50": Object {
      "value": "d",
    },
    "q51": Object {
      "value": "a",
    },
    "q52": Object {
      "value": "b",
    },
    "q53": Object {
      "value": "b",
    },
    "q54": Object {
      "value": "d",
    },
    "q55": Object {
      "value": "*",
    },
    "q56": Object {
      "value": "b",
    },
    "q57": Object {
      "value": "a",
    },
    "q58": Object {
      "value": "d",
    },
    "q59": Object {
      "value": "a",
    },
    "q6": Object {
      "value": "d",
    },
    "q60": Object {
      "value": "a",
    },
    "q7": Object {
      "value": "c",
    },
    "q8": Object {
      "value": "a",
    },
    "q9": Object {
      "value": "d",
    },
  },
  "correctCount": 0,
  "id": Any<String>,
  "imageFile": Any<String>,
  "incorrectCount": 0,
  "isCompiled": false,
  "isRollNoExtracted": true,
  "marks": 0,
  "post": "",
  "questionPaperType": "",
  "rollNo": "10025",
  "skippedCount": 0,
  "testCenter": "",
  "testTime": "",
  "totalMarks": 0,
  "unattemptedCount": 0,
}
`;

exports[`workerExtract should be able to extract result just fine 3`] = `
Object {
  "answers": Object {
    "q1": Object {
      "value": "d",
    },
    "q10": Object {
      "value": "d",
    },
    "q11": Object {
      "value": "c",
    },
    "q12": Object {
      "value": "d",
    },
    "q13": Object {
      "value": "c",
    },
    "q14": Object {
      "value": "b",
    },
    "q15": Object {
      "value": "a",
    },
    "q16": Object {
      "value": "b",
    },
    "q17": Object {
      "value": "c",
    },
    "q18": Object {
      "value": "b",
    },
    "q19": Object {
      "value": "c",
    },
    "q2": Object {
      "value": "d",
    },
    "q20": Object {
      "value": "a",
    },
    "q21": Object {
      "value": "d",
    },
    "q22": Object {
      "value": "d",
    },
    "q23": Object {
      "value": "?",
    },
    "q24": Object {
      "value": "d",
    },
    "q25": Object {
      "value": "d",
    },
    "q26": Object {
      "value": "c",
    },
    "q27": Object {
      "value": "c",
    },
    "q28": Object {
      "value": "a",
    },
    "q29": Object {
      "value": "b",
    },
    "q3": Object {
      "value": "c",
    },
    "q30": Object {
      "value": "c",
    },
    "q31": Object {
      "value": "b",
    },
    "q32": Object {
      "value": "a",
    },
    "q33": Object {
      "value": "d",
    },
    "q34": Object {
      "value": "b",
    },
    "q35": Object {
      "value": "a",
    },
    "q36": Object {
      "value": "b",
    },
    "q37": Object {
      "value": "c",
    },
    "q38": Object {
      "value": "d",
    },
    "q39": Object {
      "value": "a",
    },
    "q4": Object {
      "value": "a",
    },
    "q40": Object {
      "value": "b",
    },
    "q41": Object {
      "value": "c",
    },
    "q42": Object {
      "value": "a",
    },
    "q43": Object {
      "value": "a",
    },
    "q44": Object {
      "value": "c",
    },
    "q45": Object {
      "value": "b",
    },
    "q46": Object {
      "value": "d",
    },
    "q47": Object {
      "value": "a",
    },
    "q48": Object {
      "value": "c",
    },
    "q49": Object {
      "value": "a",
    },
    "q5": Object {
      "value": "c",
    },
    "q50": Object {
      "value": "d",
    },
    "q51": Object {
      "value": "d",
    },
    "q52": Object {
      "value": "b",
    },
    "q53": Object {
      "value": "b",
    },
    "q54": Object {
      "value": "d",
    },
    "q55": Object {
      "value": "d",
    },
    "q56": Object {
      "value": "d",
    },
    "q57": Object {
      "value": "b",
    },
    "q58": Object {
      "value": "b",
    },
    "q59": Object {
      "value": "d",
    },
    "q6": Object {
      "value": "d",
    },
    "q60": Object {
      "value": "d",
    },
    "q7": Object {
      "value": "c",
    },
    "q8": Object {
      "value": "a",
    },
    "q9": Object {
      "value": "c",
    },
  },
  "correctCount": 0,
  "id": Any<String>,
  "imageFile": Any<String>,
  "incorrectCount": 0,
  "isCompiled": false,
  "isRollNoExtracted": false,
  "marks": 0,
  "post": "",
  "questionPaperType": "",
  "rollNo": undefined,
  "skippedCount": 0,
  "testCenter": "",
  "testTime": "",
  "totalMarks": 0,
  "unattemptedCount": 0,
}
`;

exports[`workerExtract should be able to extract result just fine 4`] = `
Object {
  "answers": Object {
    "q1": Object {
      "value": "d",
    },
    "q10": Object {
      "value": "d",
    },
    "q11": Object {
      "value": "c",
    },
    "q12": Object {
      "value": "d",
    },
    "q13": Object {
      "value": "c",
    },
    "q14": Object {
      "value": "b",
    },
    "q15": Object {
      "value": "a",
    },
    "q16": Object {
      "value": "b",
    },
    "q17": Object {
      "value": "c",
    },
    "q18": Object {
      "value": "b",
    },
    "q19": Object {
      "value": "c",
    },
    "q2": Object {
      "value": "d",
    },
    "q20": Object {
      "value": "a",
    },
    "q21": Object {
      "value": "d",
    },
    "q22": Object {
      "value": "d",
    },
    "q23": Object {
      "value": "?",
    },
    "q24": Object {
      "value": "d",
    },
    "q25": Object {
      "value": "d",
    },
    "q26": Object {
      "value": "c",
    },
    "q27": Object {
      "value": "c",
    },
    "q28": Object {
      "value": "a",
    },
    "q29": Object {
      "value": "b",
    },
    "q3": Object {
      "value": "c",
    },
    "q30": Object {
      "value": "c",
    },
    "q31": Object {
      "value": "b",
    },
    "q32": Object {
      "value": "a",
    },
    "q33": Object {
      "value": "d",
    },
    "q34": Object {
      "value": "b",
    },
    "q35": Object {
      "value": "a",
    },
    "q36": Object {
      "value": "b",
    },
    "q37": Object {
      "value": "c",
    },
    "q38": Object {
      "value": "d",
    },
    "q39": Object {
      "value": "a",
    },
    "q4": Object {
      "value": "a",
    },
    "q40": Object {
      "value": "b",
    },
    "q41": Object {
      "value": "c",
    },
    "q42": Object {
      "value": "a",
    },
    "q43": Object {
      "value": "a",
    },
    "q44": Object {
      "value": "c",
    },
    "q45": Object {
      "value": "b",
    },
    "q46": Object {
      "value": "d",
    },
    "q47": Object {
      "value": "a",
    },
    "q48": Object {
      "value": "c",
    },
    "q49": Object {
      "value": "a",
    },
    "q5": Object {
      "value": "c",
    },
    "q50": Object {
      "value": "d",
    },
    "q51": Object {
      "value": "d",
    },
    "q52": Object {
      "value": "b",
    },
    "q53": Object {
      "value": "b",
    },
    "q54": Object {
      "value": "d",
    },
    "q55": Object {
      "value": "d",
    },
    "q56": Object {
      "value": "d",
    },
    "q57": Object {
      "value": "b",
    },
    "q58": Object {
      "value": "b",
    },
    "q59": Object {
      "value": "d",
    },
    "q6": Object {
      "value": "d",
    },
    "q60": Object {
      "value": "d",
    },
    "q7": Object {
      "value": "c",
    },
    "q8": Object {
      "value": "a",
    },
    "q9": Object {
      "value": "c",
    },
  },
  "correctCount": 0,
  "id": Any<String>,
  "imageFile": Any<String>,
  "incorrectCount": 0,
  "isCompiled": false,
  "isRollNoExtracted": false,
  "marks": 0,
  "post": "",
  "questionPaperType": "",
  "rollNo": undefined,
  "skippedCount": 0,
  "testCenter": "",
  "testTime": "",
  "totalMarks": 0,
  "unattemptedCount": 0,
}
`;
